src/uct/ib/base/ib_md.c

```diff
@@ -1678,6 +1678,18 @@ void uct_ib_md_close(uct_md_h uct_md)
     ucs_free(md);
 }
 
+static void uct_ib_verbs_md_set_access_flags(uct_ib_md_t *md)
```

Comment:
But access flags and memory registration are associated with an md, not an iface/tl. For instance, `uct_ib_mem_reg()` gets a `uct_md_h` handle only. So, I'm not sure how I can set different access flags based on iface/tl.

Reply:
We should pass down flags to uct_ib_mem_reg, see uct_md_mem_flags. Should pass UCT_MD_MEM_ACCESS_LOCAL_READ|UCT_MD_MEM_ACCESS_LOCAL_WRITE

Reply:
Couldn't figure this out. I can use `uct_md_mem_flags` for memory that is allocated and registered in `uct_iface_mem_alloc()` (commit https://github.com/openucx/ucx/pull/6353/commits/4cebd30a80ad88e7983c708454432acb07e7a861). But, I can't control the flags that are passed to `uct_md_mem_reg()`. For instance, UCP always uses `UCT_MD_MEM_ACCESS_RMA` to register user buffers.
https://github.com/openucx/ucx/blob/7dd283393ad15b75a4e57b98645dc5227fbc4702/src/ucp/core/ucp_request.c#L229
We don't register memory per tl (or communication protocol)

Reply:
what is the blacktrace of failure w/o this change? is it for user buffer registration or for memory pool?
can we change the flags in uct_iface_mp_chunk_alloc -> uct_iface_mem_alloc?

Reply:
Without commit https://github.com/openucx/ucx/pull/6353/commits/4cebd30a80ad88e7983c708454432acb07e7a861 the failure happens early on for iface memory pool. But, even with commit https://github.com/openucx/ucx/pull/6353/commits/4cebd30a80ad88e7983c708454432acb07e7a861 user buffer registration will still fail

Reply:
Another problem is that some EFA devices support IBV_ACCESS_REMOTE_READ and some don't. So, I cannot set the access flags based on device type either.

Reply:
UD should not really need remote read. we need to narrow down the flags to what is really needed for the datapath operatons.

Reply:
Right, I'll see what needs to be changed at UCP layer for this.

Reply:
I updated some functions like `ucp_request_send_buffer_reg_lane()` to set the access flags based on the specific communication protocol (commit https://github.com/openucx/ucx/pull/6353/commits/e969ded142ee93b23088ce7b12e0ac32461eab16).
But I'm not sure what we should do for UCP registered memory pool if we want to narrow down the flags.
https://github.com/openucx/ucx/blob/7dd283393ad15b75a4e57b98645dc5227fbc4702/src/ucp/core/ucp_worker.c#L1777

Reply:
@yosefe This will disrupt rcache too, because rcache does not know about UCT-level access flags; it only checks for READ/WRITE protection flags. So, if a region is registered with LOCAL_READ | LOCAL_WRITE at some point, and later on we need it registered with REMOTE_READ too, we can't do it because we will get cache hit.
What is the problem with always registering with all the access flags that are supported by the device?

Reply:
the right thing would be to register with flags according to the operation we want to do on the memory
we can pass more flags, if needed, by the user defined arg to rcache_get():
```
 * @param [in]  arg         Custom argument passed down to memory registration
 *                          callback, if a memory registration happens during
 *                          this call.
 * ```

Reply:
IIUC, that `arg` is already used to pass down the access flags to the registration callback. Once registered, the access flags are stored in the memory handle. But, `ucs_rcache_get()` cannot use that `arg`, so upon a hit, it cannot check if it is a valid hit with respect to access flags. 
`uct_ib_mem_rcache_reg()` can do that check, and force a re-registration if the access flags of the memory handle do not satisfy those of the `arg`. For that, rcache needs to provide a function for force re-registration.

Reply:
@SeyedMir what if we just replace MD_ACCESS_ALL in https://github.com/openucx/ucx/pull/6353/files#diff-4d9f40ed5ed92d7678c3da565230cac1069f58245a8ebd066af86939f98896efL338 by using "flags" parameter?

Reply:
also, in order to accomplish "register with all supported access flag", I would expect something like just modifying uct_ib_md_access_flags() to check RC transport support and use IBV_LOCAL_xx flags instead of UCT_IB_MEM_ACCESS_FLAGS, and not add anything to uct_base_iface_t

Reply:
It won't be enough because that will only solve the problem for the memory that is allocated by an iface. Still, there are other places in UCP where we should decide what access flags should be passed for registration. One example is the registered mpool in UCP worker.

Reply:
AFAIK, ucp mpool should still go through uct_ib_md_access_flags to resolve access flags, isn't this so?

Reply:
That's right, and currently that function sets access flags to ACCESS_ALL regardless of what has been passed down.
In the last two commits, I updated that function so it uses the flags that have been passed down to it. But it will still fail since UCP passes ACCESS_RMA/ACCESS_ALL in many places.
That's why I wanted to store in ib md the access flags that are actually supported by the underlying device, so that in that function we use the supported flags (or supported_flags & flags).

Reply:
I think we can check if RC is supported and if not set ib_md::access_flags to LOCAL_READ|LOCAL_WRITE only. 
assuming that if no RC -> no RDMA

Reply:
It will work for this PR, but it will fail as soon as we add SRD. Because EFA does not have RC, but it will have RDMA READ which can (and will be) used once we add support for SRD.

Reply:
Is EFA exposing Verbs API?

Reply:
@SeyedMir how would you suggest to check if REMOTE_READ is supported by the device?

Reply:
I don't see any other way than actually trying to register memory and see if it fails or not. I could not find anything in device attributes which will tell what flags are supported by a device.

Reply:
how would you know if SRD is supported?
can an SRD-capable device override and add LOCAL_READ flags to the MD?

Reply:
All EFA devices will have SRD, but not all of them will have support for REMOTE_READ access flag.

Reply:
check this?
https://github.com/linux-rdma/rdma-core/blob/master/providers/efa/efadv.h#L39

Reply:
yes, I think I can query for EFA-specific attributes during device initialization once the device type is detected as EFA.

Reply:
@yosefe I updated things accordingly. Can you please take another look at this?